buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java

```diff
@@ -0,0 +1,125 @@
+                    templateFile.getParentFile().mkdirs();
+                    try {
+                        System.out.println(templateFile); ;
+                        Files.copy(resourceTemplate, templateFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+        dependsOn(copyPluginPropertiesTemplate);
+        extension = project.getExtensions().create("esplugin", PluginPropertiesExtension.class, project);
+        project.afterEvaluate((unused) -> {
+            // check require properties are set
+            if (this.extension.getName() == null) {
+                throw new InvalidUserDataException("name is a required setting for esplugin");
+            }
+            if (this.extension.getDescription() == null) {
+                throw new InvalidUserDataException("description is a required setting for esplugin");
+            }
+            if (this.extension.getClassname() == null) {
+                throw new InvalidUserDataException("classname is a required setting for esplugin");
+            }
+
+            from(templateFile.getParentFile()).include(this.descriptorOutput.getName());
+            into(this.descriptorOutput.getParentFile());
+            Map<String, String> properties = this.generateSubstitutions();
+            expand(properties);
+            this.getInputs().properties(properties);
+        });
+
+    }
+
+    private Map<String, String> generateSubstitutions() {
+        Map<String, String> substitutions = new HashMap<>();
+        substitutions.put("name", this.extension.getName());
+        substitutions.put("description", this.extension.getDescription());
+        substitutions.put("version", this.extension.getVersion());
+        substitutions.put("elasticsearchVersion", Version.fromString(VersionProperties.getElasticsearch()).toString());
+
+        //TODO (Raf): No getter for project.getTargetCompatibility()? What's the alternative option for below line?
+        substitutions.put("javaVersion", String.valueOf(this.getProject().getVersion()));
+        substitutions.put("classname", this.extension.getClassname());
+
+        //TODO (Raf): Should we worry about possible NPE on getExtendedPlugins() here?
+        substitutions.put("extendedPlugins", this.extension.getExtendedPlugins().stream().collect(Collectors.joining(",")));
+        substitutions.put("hasNativeController", String.valueOf(this.extension.isHasNativeController()));
+        substitutions.put("requiresKeystore", String.valueOf(this.extension.isRequiresKeystore()));
+
+        return substitutions;
+    }
+
+    @Input
+    public PluginPropertiesExtension getExtension() {
+        return extension;
+    }
+
+    public void setExtension(PluginPropertiesExtension extension) {
```

Comment:
I don't think we should have a setter for it, the task should discover it from project.